packages/itwin/tree-widget/src/test/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.test.ts

```diff
@@ -2193,6 +2204,409 @@ describe("HierarchyBasedVisibilityHandler", () => {
         });
       });
     });
+
+    describe("filtered nodes", async () => {
```

Comment:
I trust these tests do exercise the situations, but I personally find it very hard to understand what they do or what I'd need to do to add an additional test for a specific case. This is because:

- Use of a single iModel for all test cases. That's good for performance, but makes maintaining the tests way more difficult, as you constantly have to update old tests when adding new test cases that require changes to the dataset. In addition, because the number of different instances in the iModel becomes unstable in the context of a single test, you have to take pretty much random instance of each type (subject, model, category, geometric element) when creating key paths and selecting which one to use for simulating a visibility change.

- Use of heavily parametrized test functions. This helps avoid code duplication, but makes the tests hard to read and maintain - parametrized functions become more complex every time you need to add an additional test for a special case.

My suggestion would be to restructure the tests in this form:
1. Create individual iModel with instances required for the specific test case.
2. Create filtered instance key paths from the created instance keys.
3. Call a function similar to `validateHierarchy`, where you specify a provider and a hierarchy of instance key + expected visibility. You can call it twice - for filtered and unfiltered hierarchy.

Reply:
Rewrote all integration tests using the pattern similar to what you've suggested.

Reply:
definitely better now, thanks!